### ISSUE TITLE
pyFFTW breaks when forked

### ISSUE THREAD

**Issue by robbmcleod** [hgomersall/pyFFTW#65]
Hi Homer and others,

I've been building some code that mixes together pyfftw and numexpr, and I want to run multiple jobs on a single machine. In general I want to do the plans ahead of time both for efficiency but also because letting the builder helper module do the planning while the machine is under load results in some truly terrible wisdom.  It seems `pyfftw` has some trouble if it's forked by the Python `multiprocessing` library if `pyfftw` has been touched at all.  Here's some example code:

    import multiprocessing as mp
    import pyfftw, pickle, time
    import numpy as np
    
    def mpFFTW( processID ):
        # reload( pyfftw ) # This crashes...
        
        print( "MP importing wisdom: " + str(processID) )
        fh = open( pickle_name, 'rb')
        pyfftw.import_wisdom( pickle.load( fh ) )
        fh.close()
        
        print( "MP Setup: " + str(processID) )
        temp1 = pyfftw.n_byte_align_empty( shapePadded, fftwdtype.itemsize, dtype=fftwdtype )
        FFT2 = pyfftw.builders.fft2( temp1, planner_effort=fftw_effort , threads=n_threads )
        rand1 = np.random.normal( size=shapePadded ).astype('complex64')
        rand2 = np.zeros( shapePadded, dtype='complex64' )
        FFT2.update_arrays( rand1, rand2 )
        print( "Execution: " + str(processID) )
        FFT2.execute()
        print( "Finished: " + str(processID) ) # We never get here
        pass
        
    # Take the list of filenames and split it up into mp_processes pieces
    # Windows multiprocessing protection:
    if __name__ == "__main__":
        pickle_name = "./pyfftw_wisdom.pkl"# import threading as mp
        mp_processes = 4
        over_subscribe = 1.0 
        sleep_time = 2 # Offset time between starting each process for disk IO
        fftw_effort = 'FFTW_MEASURE'
        shapePadded = [2048,2048]
        
        n_threads = np.int( mp.cpu_count() / mp_processes * over_subscribe ) # 12 for example
        
        # Make a plan for FFTW
        print( "Planning FFTW for Multiprocessing" )
        try:
            fh = open( pickle_name, 'wb')
            # Start from scratch
            pyfftw.forget_wisdom()
            
            fftwdtype = np.dtype('complex64') # I am assuming input data is float32
            # full size
            temp1 = pyfftw.n_byte_align_empty( shapePadded, fftwdtype.itemsize, dtype=fftwdtype )
            temp2 = pyfftw.n_byte_align_empty( shapePadded, fftwdtype.itemsize, dtype=fftwdtype )
            FFT2 = pyfftw.builders.fft2( temp1, planner_effort=fftw_effort , threads=n_threads )
            FFT2.update_arrays( temp1, temp2 ); FFT2.execute()
    
            fftw_plan = pyfftw.export_wisdom()    
            print( "FFTW wisdom done" )
            pickle.dump( fftw_plan, fh )
            fh.close()
        except IOError:
            raise IOError( "Failed to open FFTW wisdom pickle file: " + pickle_name )
            
        # Test the built FFTW object
        print( "Planner testing" )
        rand1 = np.random.normal( size=[2048,2048] ).astype('complex64')
        rand2 = np.zeros( [2048,2048], dtype='complex64' )
        FFT2.update_arrays( rand1, rand2 )
        print( "Execution " )
        FFT2.execute() # Works
        
        # Launch individual processes
        procs = []
        for J in xrange(0,mp_processes):
            print( "Launching process: " + str(J) )
            procs.append( mp.Process( target=mpFFTW, args=( str(J), ) ) )
            procs[J].daemon = True
            procs[J].start()
            time.sleep(sleep_time)
        
        # Join waits forever...
        for proc in procs:
            proc.join()
        print( "All processes successfully finished" )

The child processes never seem to finish their calls to FFT2.execute().  Looking at the code the only thing that jumps out is that it may be stuck in the GIL?

There is a work-around, don't call any pyfftw functions from the parent process.  This means doing the wisdom planning in a separate child process, saving the wisdom file to disk, and wait for the child to join, before launching the worker processes.

Also, I noticed when debugging that I can't call reload(pyfftw).

**Comment by hgomersall** [comment]
I have very little experience with `multiprocessor` (the only experience I have is that it seems to break everything!).

I understand it sort of forks the whole memory space, but with some copy-on-write optimisations. I don't know if this is an FFTW issue - have you tried forking processes that do a similar thing with the C library?

It's not clear to me what you're actually trying to achieve here. Are you able to elaborate?

**Comment by robbmcleod** [comment]
Homer,

I haven't tried calling C-FFTW directly. I haven't tested but this is probably unique to Linux, Windows can't pass this sort of stuff to child processes.  I expect it would be ok if I could break the link to pyfftw within the subprocesses and make a new copy of the FFTW .dlls in memory for each child process, but like I said in a comment in the code, I can't call `rebuild(pyfftw)` as it throws an error:

    File "/mnt/ssd/mpImportTest.py", line 19, in mpFFTW
    reload( pyfftw ) # This crashes...
    File "/mnt/ssd/anaconda/lib/python2.7/site-packages/pyfftw/__init__.py", line 30, in <module>
    del builders.builders
    AttributeError: builders

I don't know if there's another way in Python to refresh a library?  Looking on Stack Exchange it seems impossible:

http://stackoverflow.com/questions/3105801/unload-a-module-in-python

It's not the end of the world.  It does imply I need to firewall each process that calls pyFFTW, so any preprocessing shared between processes needs to be saved to a file (which may be in RAMdisk or shared memory region or whatever).  

To elaborate and provide more background : I am writing a library that does image alignment by cross-correlations on image stacks.  However to improve robustness we use many templates (potentially we cross-correlate every image to every other image).  So we might need to do 1000 cross-correlation operations, on 4k x 4k images, and each correlation operation is a bunch of matrix operations and 5-7 Fourier transforms.  It's a relatively heavy task for a single machine.  To improve performance, we want to run multiple copies both on the same machine, and other machines in a small cluster. Both pyfftw and numexpr provide blocking and multithreaded operations (for Fourier transforms and matrix operations respectively). However there's still a fair amount of CPU time that is idle when we are in single-threaded Python.  Therefore to try to improve utilization, I run 2-4 processes in parallel on big directories of image stacks. Currently I can sustain 80 % CPU load on 48 virtual cores, so this approach does work.

**Comment by hgomersall** [comment]
Hi Rob,

It's Henry (not Homer :)

So I actually got around to testing your code. It works just fine on my machine. Am I missing something here? Are you using pyFFTW from git?

Additionally: there has been thought given to supporting MPI with pyfftw (see #38). I'm sure such an implementation would be well received and I'd be happy to advise in any way I can (I just don't have much time at the moment to commit to working on such things, it being of marginal interest to me personally). Perhaps PFFT, referenced in #38 would be useful here?

**Comment by robbmcleod** [comment]
Sorry about the name, my brain skips right over your username and melds it into one.

I'm using pyFFTW cloned from git.  On windows machines, and on Python 3.4 on linux it should not be a problem (or at least it can be fixed).  It is a problem on legacy Linux versions.  Did you really get to the 
    print( "All processes successfully finished" )
line?  For me I get stuck in the FFT2.execute() calls with Anaconda Python 2.7.  If you are finishing I may have to look into how FFTW was compiled.

I tried a couple of things but it seems `reload(pyfftw)` does not solve my problem. Apparently forking in linux does not create new copies of shared libraries in memory for each child:

http://stackoverflow.com/questions/29735452/when-a-process-forks-would-the-shared-library-so-still-in-the-address-space-a

I'll try looking at the MPI libraries at some point but I'm not sure there's a good solution. I'm taking a look at the Cython mailing list archives at the moment.

**Comment by hgomersall** [comment]
Yes, my output is:

```
Planning FFTW for Multiprocessing
FFTW wisdom done
Planner testing
Execution 
Launching process: 0
MP importing wisdom: 0
MP Setup: 0
Execution: 0
Finished: 0
Launching process: 1
MP importing wisdom: 1
MP Setup: 1
Execution: 1
Finished: 1
Launching process: 2
MP importing wisdom: 2
MP Setup: 2
Execution: 2
Finished: 2
Launching process: 3
MP importing wisdom: 3
MP Setup: 3
Execution: 3
Finished: 3
All processes successfully finished
```

**Comment by hgomersall** [comment]
Sorry, to clarify, that is with 2.7 under Linux, with FFTW installed from the ubuntu repository.

**Comment by hgomersall** [comment]
Rob, any movement on this? What python version are you running to show this bug?

**Comment by robbmcleod** [comment]
Hi Henry, sorry I had ortho surgery recently and forgot about it.  The problem seems to have gone away after a recent Ubuntu kernel update (and in general FFTW got faster).  So it was either a kernel problem, or more likely a compilation problem with FFTW itself.

**Comment by hgomersall** [comment]
The best kind of bug, a non existent one :)
